odoo/tools/func.py

```diff
@@ -1,25 +1,42 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
+from __future__ import annotations
+import typing
+from inspect import Parameter, getsourcefile, signature
 
-__all__ = ['synchronized', 'lazy_classproperty', 'lazy_property',
-           'classproperty', 'conditional', 'lazy']
+from decorator import decorator
 
-from inspect import getsourcefile, Parameter, signature
+__all__ = [
+    'classproperty',
+    'conditional',
+    'lazy',
+    'lazy_classproperty',
+    'lazy_property',
+    'synchronized',
+]
 
-from decorator import decorator
+T = typing.TypeVar("T")
+
+if typing.TYPE_CHECKING:
+    from collections.abc import Callable
 
-class lazy_property(object):
+
+class lazy_property(typing.Generic[T]):
     """ Decorator for a lazy property of an object, i.e., an object attribute
         that is determined by the result of a method call evaluated once. To
         reevaluate the property, simply delete the attribute on the object, and
         get it again.
     """
-    def __init__(self, fget):
+    def __init__(self, fget: Callable[[typing.Any], T]):
         assert not fget.__name__.startswith('__'),\
             "lazy_property does not support mangled names"
         self.fget = fget
 
-    def __get__(self, obj, cls):
+    @typing.overload
+    def __get__(self, obj: None, cls: typing.Any, /) -> typing.Any: ...
+    @typing.overload
+    def __get__(self, obj: object, cls: typing.Any, /) -> T: ...
+
+    def __get__(self, obj, cls, /):
```

Comment:
I don't know the symbol `/`, what is it for?

Reply:
All arguments preceding cannot be passed by keyword. This ensures the method signatures are equivalent and you can use meaningful names. Also it makes no sense to call these with kw values and it's slightly more performant.
Ref. [PEP-570](https://peps.python.org/pep-0570/)